include/deal.II/lac/dynamic_sparsity_pattern.h

```diff
@@ -732,19 +772,23 @@ namespace DynamicSparsityPatternIterators
     // if this moves us beyond the end of the row, go to the next row
     // if possible, or set the iterator to an invalid state if not.
     //
-    // use a while loop to ensure that we properly skip over lines
-    // that are empty
-    while (current_entry ==
-           sparsity_pattern->lines[current_row].entries.end())
+    // going to the next row is a bit complicated because we may have
+    // to skip over empty rows, and because we also have to avoid rows
+    // that aren't listed in a possibly passed IndexSet argument of
+    // the sparsity pattern. consequently, rather than trying to
+    // duplicate code here, just call the begin() function of the
+    // sparsity pattern itself
+    if (current_entry ==
+        ((sparsity_pattern->rowset.size()==0)
+         ?
+         sparsity_pattern->lines[current_row].entries.end()
+         :
+         sparsity_pattern->lines[sparsity_pattern->rowset.index_within_set(current_row)].entries.end()))
```

Comment:
This looks clearly more expensive. Without having looked into it more deeply, would it be easier to run over the local lines only? Then hold that value as `current_row` and make a index computation for the `row()` access function. I think this would be cheaper for advance() which is the method that really matters.